core/trino-main/src/main/java/io/trino/metadata/DiscoveryNodeManager.java

```diff
@@ -229,55 +229,65 @@ private synchronized void refreshNodesInternal()
                 InternalNode node = new InternalNode(service.getNodeId(), uri, nodeVersion, coordinator);
                 NodeState nodeState = getNodeState(node);
 
-                switch (nodeState) {
-                    case ACTIVE:
-                        activeNodesBuilder.add(node);
-                        if (coordinator) {
-                            coordinatorsBuilder.add(node);
-                        }
+                // nodesToBeDecommissioned is the authoritative list of node to be decommissioned
+                // from coordinator perspective. Once a worker appears in the list,
+                // its state become DECOMMISSIONING even if worker has yet confirmed such
+                // so that no new tasks will be scheduled on it.
+                if (!coordinator && nodesToBeDecommissioned.contains(node.getNodeIdentifier())
+                        && nodeState == ACTIVE) {
+                    log.debug("Treat " + node.getNodeIdentifier() + " as DECOMMISSIONING");
+                    nodeState = DECOMMISSIONING;
+                }
```

Comment:
Honestly I am not a big fan of this dichotomy. I would rather keep what workers report as a source of truth, and report that.

Also the quesion here may become obsolete if we would decide to drop `NodesResource.refreshNodes` as discused in https://github.com/trinodb/trino/pull/14876/files#r1051663508